### PR TITLE
Reworked proximity renderer

### PR DESCRIPTION
**Pull request by favreau**
The idea is to be able to show or hide the geometry without impacting the rendering of the touches. New rendering parameters are:
- Surface shading: Activates or deactivates the rendering of the surfaces
- Alpha correction: Used to blend colors as the ray goes through intersections

### CODE REVIEW

File: tests/braynsTestData.cpp [hunk 1]
```diff
@@ -213,6 +213,31 @@ BOOST_AUTO_TEST_CASE(render_circuit_with_particle_renderer)
                                  brayns.getEngine().getFrameBuffer()));
 }
 
+BOOST_AUTO_TEST_CASE(render_circuit_with_proximity_renderer)
+{
+    auto& testSuite = boost::unit_test::framework::master_test_suite();
+
+    const char* app = testSuite.argv[0];
+    const char* argv[] = {app,         "--renderer",
```
Comment (rolandjitsu) [inline tests/braynsTestData.cpp]: The args are formatted weird. Why not:
```cpp
const char* argv[] = {app, "demo",
                      "--renderer", "proximity",
                      "--samples-per-pixel", "16",
                      "--synchronous-mode", "true"};
```

**Comment by favreau** [comment]
@tribal-tec. Just added a description to the PR. Let me know if it's still unclear.

**Comment by tribal-tec** [comment]
Thx, much better. Please put that to the commit message also when you rebase/merge.

File: tests/braynsTestData.cpp (hunk 1 above)
Comment (favreau) [reply tests/braynsTestData.cpp]: The format applied here is what we agreed on when we decided to use clang-format. I would not see the point of making an exception here. And I don't see any issue with the new formatting, it's clear and not messy at all. Am I missing something?

File: tests/braynsTestData.cpp [hunk 2]
```diff
@@ -222,6 +222,31 @@ BOOST_AUTO_TEST_CASE(render_circuit_with_particle_renderer)
                                  brayns.getEngine().getFrameBuffer()));
 }
 
+BOOST_AUTO_TEST_CASE(render_circuit_with_proximity_renderer)
```
Comment (tribal-tec) [inline tests/braynsTestData.cpp]: This test doesn't use a circuit here. We should move all the tests that do not need BBPTestDaa to a separate file.

File: tests/braynsTestData.cpp [hunk 3]
```diff
@@ -222,6 +222,31 @@ BOOST_AUTO_TEST_CASE(render_circuit_with_particle_renderer)
...
+    const brayns::Vector3f rotCenter = {0.5f, 0.5f, 0.5f};
+
+    auto& camera = brayns.getEngine().getCamera();
+    const auto camPos = camera.getPosition();
+
+    camera.setTarget(rotCenter);
+    camera.setPosition(camPos - (rotCenter - camPos));
+
+    brayns.render();
+    BOOST_CHECK(compareTestImage("testdataallmini50proximity.png",
```
Comment (tribal-tec) [inline tests/braynsTestData.cpp]: testproximity.png only?

File: plugins/engines/ospray/ispc/render/ProximityRenderer.ispc [hunk 4]
```diff
@@ -28,178 +28,147 @@ struct ProximityRenderer
...
+            if (ao_ray.geomID >= 0)
             {
-                // Intersection detected
+                continueWithSurfaceShading = false;
+
                 postIntersect(self->super.super.model, dg, ao_ray,
                               DG_MATERIALID);
 
-                varying bool test = true;
+                bool test = true;
```
[the remark this answers is not in the document]
Comment (favreau) [reply plugins/engines/ospray/ispc/render/ProximityRenderer.ispc]: Ok

File: plugins/engines/ospray/ispc/render/ProximityRenderer.ispc [hunk 5]
```diff
@@ -28,178 +28,147 @@ struct ProximityRenderer
...
-            if (ao_ray.t != t_max)
+            if (ao_ray.geomID >= 0)
             {
-                // Intersection detected
+                continueWithSurfaceShading = false;
+
                 postIntersect(self->super.super.model, dg, ao_ray,
                               DG_MATERIALID);
 
-                varying bool test = true;
```
[the remark this answers is not in the document]
Comment (favreau) [reply plugins/engines/ospray/ispc/render/ProximityRenderer.ispc]: Ok

File: plugins/engines/ospray/ispc/render/ProximityRenderer.ispc [hunk 6]
```diff
@@ -28,178 +28,147 @@ struct ProximityRenderer
...
-                    test = material != dg.material;
+                    test = (material != dg.material);
 
                 if (test)
                 {
-                    if (ao_ray.t < t_max * 0.2f)
-                        color = self->nearColor * opacity / ao_ray.t;
+                    const float a = ao_ray.t / self->detectionDistance;
+                    const vec4f touchColor =
+                        make_vec4f(a > 0.2f ? self->nearColor : self->farColor,
```
[the remark this answers is not in the document]
Comment (favreau) [reply plugins/engines/ospray/ispc/render/ProximityRenderer.ispc]: Ok. created a nearFarThreshold constant.

File: tests/braynsTestData.cpp (hunk 3 above)
Comment (favreau) [reply tests/braynsTestData.cpp]: Ok. testDemoProximity.png

File: tests/braynsTestData.cpp (hunk 2 above)
Comment (favreau) [reply tests/braynsTestData.cpp]: I agree, but will do a separate PR for that.

File: plugins/engines/ospray/ispc/render/ProximityRenderer.ispc [hunk 7]
```diff
@@ -28,178 +28,147 @@ struct ProximityRenderer
...
-                colorKs = max(colorKs, Ks * powf(cosLR, Ns));
-
-                color = color + unshaded_light_contrib;
-            }
-        }
-
-        if (self->detectionDistance > 0.f &&
-            ray.t < self->detectionDistance * 1000.f)
+        bool continueWithSurfaceShading = true;
+        if (self->detectionDistance)
```
[the remark this answers is not in the document]
Comment (favreau) [reply plugins/engines/ospray/ispc/render/ProximityRenderer.ispc]: Hum... it's not a boolean, it really is a test on the value. We do the same for shadow intensity, and ambient occlusion strength for example. I don't think we need to introduce a new boolean variable there. Ok with you?

[the remark this answers is not in the document]
Comment (favreau) [reply plugins/engines/ospray/ispc/render/ProximityRenderer.ispc]: As discussed, forget my comment.

File: tests/braynsTestData.cpp [hunk 8]
```diff
@@ -222,6 +222,31 @@ BOOST_AUTO_TEST_CASE(render_circuit_with_particle_renderer)
...
+
+    const brayns::Vector3f rotCenter = {0.5f, 0.5f, 0.5f};
+
+    auto& camera = brayns.getEngine().getCamera();
+    const auto camPos = camera.getPosition();
+
+    camera.setTarget(rotCenter);
+    camera.setPosition(camPos - (rotCenter - camPos));
+
+    brayns.render();
```
Comment (tribal-tec) [inline tests/braynsTestData.cpp]: After my previous PR, you have to call commitAndRender() here now to fix the test error in CI.

Comment (favreau) [reply tests/braynsTestData.cpp]: Thanks for the tip.

Comment (favreau) [reply tests/braynsTestData.cpp]: hum... fails in debug. 
`Test is aborted
Leaving test module "braynsAddModelFromBlob"; testing time: 2179609us
LeakSanitizer: bad pointer 0x000000000056`

**Comment by favreau** [comment]
Retest this please

**Pull request merged**